File: servidor/fuente/RedDiff.Aplicacion/Cumplimiento/ModelosCumplimiento.cs
namespace RedDiff.Aplicacion.Cumplimiento;

public sealed record CrearReglaBaselineSolicitud(
    string Criterio,
    string Esperado,
    bool Obligatoria);

public sealed record CrearBaselineSolicitud(
    string Nombre,
    string Alcance,
    long? DispositivoId,
    string? TipoDispositivo,
    long? VersionReferenciaId,
    IReadOnlyList<CrearReglaBaselineSolicitud> Reglas);

public sealed record CambiarEstadoBaselineSolicitud(bool Activa);

public sealed record ReglaBaselineResumen(
    long Id,
    string Criterio,
    string Esperado,
    bool Obligatoria);

public sealed record VersionReferenciaBaselineResumen(
    long Id,
    int Numero,
    string Dispositivo,
    string Hash,
    DateTimeOffset CapturadaEn,
    bool Estable);

public sealed record BaselineResumen(
    long Id,
    string Nombre,
    string Alcance,
    long? DispositivoId,
    string? Dispositivo,
    string? TipoDispositivo,
    bool Activa,
    VersionReferenciaBaselineResumen? VersionReferencia,
    IReadOnlyList<ReglaBaselineResumen> Reglas);

public sealed record VerificarVersionSolicitud(
    long BaselineId,
    long VersionId);

public sealed record ResultadoReglaResumen(
    long ReglaId,
    string Criterio,
    string Esperado,
    bool Obligatoria,
    string Estado,
    string Evidencia);

public sealed record VerificacionResumen(
    long Id,
    long BaselineId,
    string Baseline,
    long DispositivoId,
    string Dispositivo,
    long VersionId,
    int Version,
    string Usuario,
    DateTimeOffset Fecha,
    string Estado,
    string ResultadoGeneral,
    int TotalReglas,
    int Cumplidas,
    int Incumplidas,
    int NoEvaluables,
    decimal PorcentajeCumplimiento);

public sealed record VerificacionDetalle(
    long Id,
    long BaselineId,
    string Baseline,
    long DispositivoId,
    string Dispositivo,
    long VersionId,
    int Version,
    string Usuario,
    DateTimeOffset Fecha,
    string Estado,
    string ResultadoGeneral,
    int TotalReglas,
    int Cumplidas,
    int Incumplidas,
    int NoEvaluables,
    decimal PorcentajeCumplimiento,
    IReadOnlyList<ResultadoReglaResumen> Resultados);

public sealed record ResultadoEvaluacionRegla(
    string Estado,
    string Evidencia);

File: cliente/src/app/modulos/cumplimiento/modelos-cumplimiento.ts
export type AlcanceBaseline = 'Dispositivo' | 'TipoDispositivo';

export type CriterioReglaBaseline = 'Contiene' | 'NoContiene' | 'CoincideExpresionRegular';

export type EstadoResultadoRegla = 'Cumplida' | 'Incumplida' | 'NoEvaluable';

export type ResultadoGeneralVerificacion = 'Cumple' | 'Incumple' | 'NoEvaluable';

export interface CrearReglaBaselineSolicitud {
  criterio: CriterioReglaBaseline;
  esperado: string;
  obligatoria: boolean;
}

export interface CrearBaselineSolicitud {
  nombre: string;
  alcance: AlcanceBaseline;
  dispositivoId: number | null;
  tipoDispositivo: string | null;
  versionReferenciaId: number | null;
  reglas: CrearReglaBaselineSolicitud[];
}

export interface ReglaBaselineResumen {
  id: number;
  criterio: CriterioReglaBaseline;
  esperado: string;
  obligatoria: boolean;
}

export interface VersionReferenciaBaselineResumen {
  id: number;
  numero: number;
  dispositivo: string;
  hash: string;
  capturadaEn: string;
  estable: boolean;
}

export interface BaselineResumen {
  id: number;
  nombre: string;
  alcance: AlcanceBaseline;
  dispositivoId: number | null;
  dispositivo: string | null;
  tipoDispositivo: string | null;
  activa: boolean;
  versionReferencia: VersionReferenciaBaselineResumen | null;
  reglas: ReglaBaselineResumen[];
}

export interface ResultadoReglaResumen {
  reglaId: number;
  criterio: CriterioReglaBaseline;
  esperado: string;
  obligatoria: boolean;
  estado: EstadoResultadoRegla;
  evidencia: string;
}

export interface VerificacionResumen {
  id: number;
  baselineId: number;
  baseline: string;
  dispositivoId: number;
  dispositivo: string;
  versionId: number;
  version: number;
  usuario: string;
  fecha: string;
  estado: 'Pendiente' | 'EnProceso' | 'Completado' | 'Fallido';
  resultadoGeneral: ResultadoGeneralVerificacion;
  totalReglas: number;
  cumplidas: number;
  incumplidas: number;
  noEvaluables: number;
  porcentajeCumplimiento: number;
}

export interface VerificacionDetalle extends VerificacionResumen {
  resultados: ResultadoReglaResumen[];
}
